src/XrdNet/XrdNetUtils.cc

```diff
@@ -938,3 +939,71 @@ bool XrdNetUtils::Singleton(const char  *hSpec, const char **eText)
 //
    return isSingle;
 }
+
+bool XrdNetUtils::ConnectWithTimeout(int sockfd, const struct sockaddr* clientAddr,size_t clientAddrLen, uint32_t timeout_sec, std::stringstream &errMsg) {
+
+  if (!SetSockBlocking(sockfd, false, errMsg)) {  // Set to non-blocking
+    close(sockfd);
+    return false;
+  }
+
+  int result = connect(sockfd, clientAddr, clientAddrLen);
+  if (result == 0) {
+    //We've managed to connect immediately
+    SetSockBlocking(sockfd, true, errMsg);  // Set the socket back to blocking
+    return true;
+  } else if (errno != EINPROGRESS) {
+    errMsg << "Connection failed: " << strerror(errno);
+    close(sockfd);
+    return false;
+  }
+
+  fd_set set;
+  struct timeval timeout;
+  FD_ZERO(&set);
+  FD_SET(sockfd, &set);
+  timeout.tv_sec = timeout_sec;
+  timeout.tv_usec = 0;
+
+  result = select(sockfd + 1, NULL, &set, NULL, &timeout);
```

Comment:
It's much better to use poll() for this and avoid all the bit vector mucking for a single fd. See XrdLinkXeq::Peek() for an example how that is done.

Reply:
OK very nice, did not know about `poll()`, will do :)

Reply:
Done in https://github.com/xrootd/xrootd/pull/2242/commits/d283708019b8d3a101979443da377f7b81b62f71
Will squash everything once everything is OK :)